File: python/segment_tree/trees/SegmentTree.py
from math import log2


class SegmentTree:

    def __init__(self, data):
        self.data = data
        self.len_data = len(data)
        self.height_tree = log2(len(data))
        self.mem_size = int(2 * 2 ** self.height_tree - 1)
        self.segment_tree = [0] * self.mem_size
        self.build_tree(0, self.len_data-1, 0)

    def get_middle(self, s, e):
        return (e - s) // 2 + s

    '''
        :param data = array para construir a segment tree
        :param s = indice que indica o inicio do array 'data'
        :param e = indice que indica o fim do array 'data'
        :param cur_index = indice da segment_tree. Usamos essa valor
        para definir qual o valor do n-esino nó da Segment Tree
    '''
    def build_tree(self, s, e, cur_index):
        if s == e:
            self.segment_tree[cur_index] = self.data[e]
            return self.data[e]
        middle_index = self.get_middle(s, e)
        '''
        usamos um array para representar a Segment Tree
        dado um indice i, _tree[i] eh o no raiz e _tree[i*2+1] eh o
        no da esquerda e _tree[i*2+2] o no da direita
        '''
        l_value = self.build_tree(s, middle_index, cur_index * 2 + 1)
        r_value = self.build_tree(middle_index + 1, e, cur_index *2 + 2)
        self.segment_tree[cur_index] = l_value + r_value
        return self.segment_tree[cur_index]


    '''
        :parameter index: indice que se quer atualizar no array 'data'
        :parameter value: valor para atualizar
    '''
    def update(self, index, value):
        if index < 0 or index > self.len_data - 1:
            return
        # valor que precisa ser adicionado aos elementos da arvore de segmento
        diff = self.data[index] - value
        self.data[index] = value
        return self._update(0, self.len_data - 1, index, diff, 0)

    def _update(self, li, ls, index, value, cur_index):
        if index < li or index > ls:
            return

        self.segment_tree[cur_index] += value
        if li != ls:
            middle_index = self.get_middle(li, ls)
            self._update(li, middle_index, index, value, cur_index * 2 + 1)
            self._update(middle_index+1, ls, index, value, cur_index * 2 + 2)

        return

    '''
        :parameter li = limite inferior da arvore (0)
        :parameter ls = limite superiro da arvore (n)
        : i e j sao os limites que se quer saber a soma
    '''
    def _prefix_sum(self, li, ls, i, j, cur_index):
        if i > ls or j < li or i > j:
            return 0
        elif i <= li and j >= ls:
            return self.segment_tree[cur_index]
        middle_index = self.get_middle(i, j)
        l_value = self._prefix_sum(li, middle_index, i, j, cur_index * 2 + 1)
        r_value = self._prefix_sum(middle_index + 1, ls, i, j, cur_index * 2 + 2)
        return l_value + r_value

    def prefix_sum(self, i, j):
        if i < 0 or j > self.len_data - 1:
            return 0
        return self._prefix_sum(0, self.len_data - 1, i, j, 0)


matrix = [
    [1, 3, 5, 7, 9, 11]
    , [1, 2, 3, 4, 5, 6, 7, 8]
]
idx = 0
st = SegmentTree(matrix[idx])

print(st.prefix_sum(0, 3))
print(st.prefix_sum(1, 3))
print(st.update(1, 10))
#print(st.update(0, 10))

if __name__ == '__main__':
    pass

File: java/SegmentTree/src/impl/SegmentTree.java
package impl;

/**
 * https://www.geeksforgeeks.org/segment-tree-set-1-range-minimum-query/
 *
 * */

public class SegmentTree {

    private static int [] tree;

    public static double log(long lgn, long base) {
        return Math.log(lgn) / Math.log(base);
    }

    private static double log2(long lgn) {
        return Math.log(lgn)/Math.log(2);
    }

    private static void build(int [] data) {
        int n = data.length;
        // altura da arvore = log2(n)
        int heightTree = (int) Math.ceil(log2(n));
        // memoria necessario = 2 * (2 ^ log2(n)) - 1
        int maxSize = (int) (2 * Math.pow(2, heightTree) - 1);
        tree = new int[maxSize];
        build(data, 0, n-1, 0);
    }

    private static int getLeft(int idx) {
        return 2 * idx + 1;
    }

    private static int getRight(int idx) {
        return getLeft(idx) + 1;
    }

    private static int getParent(int idx) {
        return (idx & 1) ==  0 ? idx / 2 - 1 : idx / 2;
    }

    /**
     * Constroi recursivamente a arvore de segmento no intervalor l ... r
     * a variavel ith eh o indice do i-esimo valor na arvore de segmento
     *
     * A forma de montar uma arvore de segmento varia conforme o problema que se quer resolver.
     * No exemplo implmentando queremos criar uma arvore que guarda a soma parcial de um array
     * dentro de um intervalor 0 <= l, r <= n (n sendo o tamanho do array)
     *
     * */
    private static int build(int data [], int l, int r, int ith) {
        if (l == r) {
            tree[ith] = data[l];        // preencher valor da folha
            return data[l];             // retornar o valor que foi adicionado a folha - esquerda ou direita
        }
        int middle = getMiddle(l, r);
        int valueLeft = build(data, l, middle, getLeft(ith));
        int valueRight = build(data, middle+1, r, getRight(ith));
        // adicionar ao no pai a soma dos filhis
        tree[ith] = valueLeft + valueRight; // preecher valor raiz = soma da folha da esq + a dir
        return tree[ith];
    }

    private static int getMiddle(int s, int e) {
        return (e - s) / 2 + s;
    }

    /**
     * Devolver a soma no intervalo l ... r dum array de tamanho n
     * */
    private static int getIntervalSum(int l, int r, int n) {
        if(l < 0 || l > n-1 || r < 0 || r > n-1 || l > r)
            return -1;
        return getIntervalSum(0, n-1, l, r, 0);
    }

    /**
     * funcao auxiliar para recuperar a soma de um intervalo l ... r
     * num array de intervalor s ... e
     * */
    private static int getIntervalSum(int s, int e, int l, int r, int ith) {
        // se um dos limites do intervalo de busca estiver fora do intervalo do array, nao da para recuperar a soma
        if (l > e || r < s)
            return 0;
        else if ( l <= s && r >= e)
            return tree[ith];
        int middle = getMiddle(s, e);
        int valueLeft = getIntervalSum(s, middle, l, r, 2 * ith + 1);
        int valueRight = getIntervalSum(middle + 1, e, l, r, 2 * ith + 2);
        return valueLeft + valueRight;
    }

    /**
     * Atualizar o valor do array na i-esima posicao e atualizar o arvore de segment em logn
     * */
    private static void update(int [] data, int size, int idxUpdate, int value) {
        // indice que deve ser atualizado esta fora dos limites da estrutura de dados
        if (idxUpdate < 0 || idxUpdate > size - 1)
            return;
        int diff = value - data[idxUpdate];
        data[idxUpdate] = value;
        update(0, size-1, idxUpdate, diff, 0);
    }

    /**
     * s = indice de onde comeca o intervalo da arvore de segmento
     * e = indice de onde termina
     * idxUpdate = indice que deve ser atualizado no array
     * diff = diferenca entre o valor antigo e o novo
     * ith = i-esimo elemento na arvore de segmento
     * */
    private static void update(int s, int e, int idxUpdate, int diff, int ith) {
        if(idxUpdate < s || idxUpdate > e)
            return;
        // atualizando a soma dos segmentos com a diferenca entre o valor antigo do array e o valor novo
        tree[ith] += diff;
        if(s != e) {
            int middle = getMiddle(s, e);
            update(s, middle, idxUpdate, diff, 2 * ith + 1);
            update(middle+1, e, idxUpdate, diff, 2 * ith  + 2);
        }
    }

    private static void test() {
        int [][] matrix = {
             {1, 3, 5, 7, 9, 11}
            ,{1, 2, 3, 4, 5, 6, 7, 8, 9, 10}
            ,{1, 2, 3, 4}
            ,{31, -2, 3, 14, 17, -1}
        };
        int [] length = {matrix[0].length
            , matrix[1].length
            , matrix[2].length
        };
        int idx = 0;
        build(matrix[idx]);
        System.out.println(getIntervalSum(0, 2, length[idx]));
        System.out.println(getIntervalSum(0, 9, length[idx]));
        update(matrix[idx], length[idx], 0, 3);
        System.out.println(getIntervalSum(0, 2, length[idx]));
        System.out.println(getIntervalSum(0, 0, length[idx]));
        System.out.println(getIntervalSum(2, 2, length[idx]));
        System.out.println(getIntervalSum(2, 5, length[idx]));
    }

    public static void main(String[] args) {
        test();
    }

}
